### PR TITLE
RFC: React Native Versions Lifecycle & Support

### PR DESCRIPTION
**Pull request by Andjeliko**
React team worked on this proposal, to help us find ways to better support React Native community in the future. 
The gist of the proposal is that we are planning to move towards supporting a limited amount of versions of React Native but support them better.

Feel free to add comments and questions! We are working on making this actionable before the end of September 2022 as it will help the team focus their efforts on supporting the most recent and the most used versions of React Native. Thank you.

### CODE REVIEW

**Comment by scarlac** [comment]
This seems reasonable to me. Only comment I have is on the "minor" wording, since it specifically refers to `y` in `x.y.z`. Since I don't know of any plans to go into 1.0 anytime soon, this isn't much a problem though. But after 1.0, wording may have to be revised slightly.

**Comment by Andjeliko** [comment]
Yes a valid point @scarlac, thanks. Also we are looking into versioning nomenclature separately.

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    * The highest version number is defined following the [NPM server algorithm](https://github.com/npm/node-semver).
+* **next version** - The next version that hasn’t been fully released yet, and is currently in development/testing (e.g. 0.69.0-RC1).
+* **minor series** - A collection of version (either stable or not stable) sharing the same minor version (e.g. **0.68**.0-RC0, **0.68**.0, **0.68**.1 are all part of the **0.68** minor series).
+
+## Which versions are currently supported?
+
+We’re supporting the **latest** **version**, and the latest versions from the **two previous minor series.**
+We’re also supporting the **next version,** having in mind this is the not yet stable version we are developing at that moment in time.
+
+Versions:
```
Comment (cipolleschi) [inline proposals/0007-rn-versions-lifecycle-support.md:28]: I think that this should be a living table, so people can refer to that when they need to upgrade or take decisions. I guess that, as soon as this proposal is approved, we would write a SLA page in the website, am I right?

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+
+|Version	|Type	|Status	|
+|---	|---	|---	|
+|0.70.0-RC2	|Release candidate	|Will be in support when stabilized	|
+|0.69.4	|Latest stable	|In support	|
+|0.68.3	|Previous minor series	|In support	|
+|0.67.4	|Previous minor series	|In support	|
```
Comment (cipolleschi) [inline proposals/0007-rn-versions-lifecycle-support.md:36]: This is also the last version with the Old Architecture. Do we want to keep supporting this version, an exceptional longer support, due to its peculiarity?

**Review by cipolleschi — APPROVED** [review]

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# RFC: React Native Versions Lifecycle & Support
+
+
+
+>As we are finding ways to better support React Native community, we are moving towards supporting a limited amount of versions of React Native better.
+
+
+The React Native Core team is working to strike a balance between the speed of innovation with the amount of regular change to the platform. We understand that each release brings new value and has an adoption cost to the community.
+This document describes how we intend to formally balance these concerns and provide clear and consistent support over time.
+React Native team will work on best-effort bases to release a new version of React Native every four months. And will focus on speeding up the release cycle and put considerable effort into ensuring a seamless upgrade path.
```
Comment (fortmarek) [inline proposals/0007-rn-versions-lifecycle-support.md:10]:
```suggestion
React Native team will work on best-effort basis to release a new version of React Native every four months. And will focus on speeding up the release cycle and put considerable effort into ensuring a seamless upgrade path.
```

Maybe instead of `React Native every four months` we could change the wording to `at least every four months`? I believe we should aim for four months to be a max, but I can also see us making release more often. That's not that clear imho from the current statement.

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+## Glossary
+
+Below we are defining terms used in this document:
+
+* **stable** **version** - Any version that doesn’t have a -alpha, -beta, -RC postfix in the version name.
+* **latest version** - **** The latest stable version with the highest version number (e.g. 0.68.2).
+    * The highest version number is defined following the [NPM server algorithm](https://github.com/npm/node-semver).
+* **next version** - The next version that hasn’t been fully released yet, and is currently in development/testing (e.g. 0.69.0-RC1).
+* **minor series** - A collection of version (either stable or not stable) sharing the same minor version (e.g. **0.68**.0-RC0, **0.68**.0, **0.68**.1 are all part of the **0.68** minor series).
```
Comment (fortmarek) [inline proposals/0007-rn-versions-lifecycle-support.md:21]:
```suggestion
* **minor series** - A collection of versions (either stable or not stable) sharing the same minor version (e.g. **0.68**.0-RC0, **0.68**.0, **0.68**.1 are all part of the **0.68** minor series).
```

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 2 above)
Comment (fortmarek) [reply proposals/0007-rn-versions-lifecycle-support.md:36]: The current versions also support the Old Architecture, is `0.67` that special?

**Review by fortmarek — APPROVED** [review]
Thanks for writing this down 🙌

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# RFC: React Native Versions Lifecycle & Support
+
+
+
+>As we are finding ways to better support React Native community, we are moving towards supporting a limited amount of versions of React Native better.
+
```
Comment (cortinico) [inline proposals/0007-rn-versions-lifecycle-support.md:6]:
```suggestion
>As we are finding ways to better support React Native community, we are moving towards supporting a limited amount of versions of React Native better.
```

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 3 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md]: Agree on this. @Andjeliko can we relax the wording?

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 1 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:28]: Yup this RFC is mostly for discussion and transparency. As we merge it, we can convert it to a page.

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 2 above)
Comment (cortinico) [inline proposals/0007-rn-versions-lifecycle-support.md:36]:
```suggestion
|0.71.0-RC0	|Release candidate	| Not released yet |
|0.70.0	|Latest stable	|In support	|
|0.69.5	|Previous minor series	|In support	|
|0.68.3	|Previous minor series	|In support	|
```

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+Versions:
+
+
+|Version	|Type	|Status	|
+|---	|---	|---	|
+|0.70.0-RC2	|Release candidate	|Will be in support when stabilized	|
+|0.69.4	|Latest stable	|In support	|
+|0.68.3	|Previous minor series	|In support	|
+|0.67.4	|Previous minor series	|In support	|
+|<=0.66.3	|Old minor series	|Unsupported	|
```
Comment (cortinico) [inline proposals/0007-rn-versions-lifecycle-support.md:37]:
```suggestion
|<=0.67.x	|Old minor series	|Unsupported	|
```

**Review by cortinico — APPROVED** [review]
We're going to keep this RFC open for a bit more to collect feedback from the community on it 👍 Feel free to share your opinion.

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 2 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:36]: Nope. Actually with the release of 0.70, React Native 0.67 went out of support.

**Comment by Andjeliko** [comment]
Thank you all for your involvement with the RFC!
We are planning to close this RFC on Monday and publish the blog post of the final agreed text on the RN blog.

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+Please note that each cherry-pick request will be assessed and approved individually. Cherry-Pick requests against unsupported versions will be rejected unless they’re involving security issues.
+
+## Security Issues
+
+Security issues should follow our issue report policy [described here](https://github.com/facebook/react-native/blob/main/CONTRIBUTING.md#security-bugs)
+
+
+## Licence
+
+React Native is open-source software distributed under the **MIT license**. Software is provided "as is" without warranty.
```
Comment (kelset) [inline proposals/0007-rn-versions-lifecycle-support.md:69]: is this section needed? 🤔

**Review by kelset — APPROVED** [review]
LGTM ✅

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+The React Native Core team is working to strike a balance between the speed of innovation with the amount of regular change to the platform. We understand that each release brings new value and has an adoption cost to the community.
+This document describes how we intend to formally balance these concerns and provide clear and consistent support over time.
+React Native team will work on best-effort basis to release a new version of React Native every four months. And will focus on speeding up the release cycle and put considerable effort into ensuring a seamless upgrade path.
+This document describes the level of support we can provide and which set of versions is supported.
+
+## Glossary
+
+Below we are defining terms used in this document:
+
+* **stable** **version** - Any version that doesn’t have a -alpha, -beta, -RC postfix in the version name.
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:17]: here we're defining stable as "what it isn't", based on a property of the release tag, rather than what "stable" means. What guarantees exist in stable (i.e. what is the stability guarantee being made?) that isn't being made for non-stable releases?

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+* **latest version** - **** The latest stable version with the highest version number (e.g. 0.68.2).
+    * The highest version number is defined following the [NPM server algorithm](https://github.com/npm/node-semver).
+* **next version** - The next version that hasn’t been fully released yet, and is currently in development/testing (e.g. 0.69.0-RC1).
+* **minor series** - A collection of versions (either stable or not stable) sharing the same minor version (e.g. **0.68**.0-RC0, **0.68**.0, **0.68**.1 are all part of the **0.68** minor series).
+
+## Which versions are currently supported?
+
+We’re supporting the **latest** **version**, and the latest versions from the **two previous minor series.**
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:25]: Can you please clarify what _supporting_ means in this context? what is the group committing to?  (this may be a pointer to a section lower below, but it's important to be precise as to what commitments are being made).

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+* Use of the New Architecture
+* Use of the Hermes Engine
+
+## Cherry-Pick Requests
+
+We’re **accepting cherry-pick requests** for the currently supported versions. A cherry-pick request is the request to include a fix in one of the supported versions and release a new point release with it.
+
+Cherry-Pick requests should be submitted via the [React Native Releases Working Group](https://github.com/reactwg/react-native-releases/discussions/categories/patches), in the corresponding discussion thread.
+
+Please note that each cherry-pick request will be assessed and approved individually. Cherry-Pick requests against unsupported versions will be rejected unless they’re involving security issues.
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:60]:
```suggestion
Please note that each cherry-pick request will be assessed and approved individually. Cherry-Pick requests against unsupported versions will be rejected unless they involve security issues.
```

[the remark this answers is not in the document]
Comment (asklar) [reply proposals/0007-rn-versions-lifecycle-support.md:60]: agree that there needs to be a tie-in to a security policy too.

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+## Security Issues
+
+Security issues should follow our issue report policy [described here](https://github.com/facebook/react-native/blob/main/CONTRIBUTING.md#security-bugs)
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:64]: this page does not mention anything related to security:
image

Comment (asklar) [reply proposals/0007-rn-versions-lifecycle-support.md:64]: in other words there is no `security-bugs` section

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+* **stable** **version** - Any version that doesn’t have a -alpha, -beta, -RC postfix in the version name.
+* **latest version** - **** The latest stable version with the highest version number (e.g. 0.68.2).
+    * The highest version number is defined following the [NPM server algorithm](https://github.com/npm/node-semver).
+* **next version** - The next version that hasn’t been fully released yet, and is currently in development/testing (e.g. 0.69.0-RC1).
+* **minor series** - A collection of versions (either stable or not stable) sharing the same minor version (e.g. **0.68**.0-RC0, **0.68**.0, **0.68**.1 are all part of the **0.68** minor series).
+
+## Which versions are currently supported?
+
+We’re supporting the **latest** **version**, and the latest versions from the **two previous minor series.**
+We’re also supporting the **next version,** having in mind this is the not yet stable version we are developing at that moment in time.
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:26]:
```suggestion
We’ll also support the **next version,** being developed, which will become stable after its release.
```

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
+# RFC: React Native Versions Lifecycle & Support
+
+
+
+>As we are finding ways to better support React Native community, we are moving towards supporting a limited amount of versions of React Native better.
+
+
+The React Native Core team is working to strike a balance between the speed of innovation with the amount of regular change to the platform. We understand that each release brings new value and has an adoption cost to the community.
+This document describes how we intend to formally balance these concerns and provide clear and consistent support over time.
+React Native team will work on best-effort basis to release a new version of React Native every four months. And will focus on speeding up the release cycle and put considerable effort into ensuring a seamless upgrade path.
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:10]: seamless upgrade: does this apply to upgrades across patch updates only, or minor releases too?

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
+# RFC: React Native Versions Lifecycle & Support
+
+
+
+>As we are finding ways to better support React Native community, we are moving towards supporting a limited amount of versions of React Native better.
```
Comment (asklar) [inline proposals/0007-rn-versions-lifecycle-support.md:5]:
```suggestion
>As we look for ways to better support the React Native community, we will move towards supporting a limited number of versions of React Native better.
```

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 7 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:69]: To some extent yes, as there are no legal obligations arising by this policy, aside from those under the MIT license (so it's like a reminder).

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 11 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:64]: Yup the link should be updated to: https://reactnative.dev/contributing/overview#security-bugs

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:60]: There is a paragraph at the end of the policy around this.
Plus there is a paragraph on the website about it: https://reactnative.dev/contributing/overview#security-bugs

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+Issues & PRs opened against older versions would be considered only in exceptional cases. Please update your application to one of the supported versions and raise the Issue/PR targeted to that version.
+
+Issues should contain a [**reproducer**](https://stackoverflow.com/help/minimal-reproducible-example) project regardless of which version they targeting, for them to be considered.
+Issues without a reproducer will require more effort to understand and fix, and are less likely to receive attention.
+
+At this point in time, we are prioritizing issues that are related to:
+
+* Latest version of React Native and two previous minor series.
+* Use of the New Architecture
+* Use of the Hermes Engine
```
[the remark this answers is not in the document]
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:52]: What do you mean? We don't expect the list to change that often.

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+## Glossary
+
+Below we are defining terms used in this document:
+
+* **stable** **version** - Any version that doesn’t have a -alpha, -beta, -RC postfix in the version name.
+* **latest version** - **** The latest stable version with the highest version number (e.g. 0.68.2).
+    * The highest version number is defined following the [NPM server algorithm](https://github.com/npm/node-semver).
+* **next version** - The next version that hasn’t been fully released yet, and is currently in development/testing (e.g. 0.69.0-RC1).
+* **minor series** - A collection of versions (either stable or not stable) sharing the same minor version (e.g. **0.68**.0-RC0, **0.68**.0, **0.68**.1 are all part of the **0.68** minor series).
```
[the remark this answers is not in the document]
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:21]: `nightly` releases are part of our testing/release infrastructure. IMHO they should not be included here as nightlies can break overnight and we can ensure they're all working (nor we can bugfix a nightly or so)

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 8 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:17]: Thanks for your suggestion.

> here we're defining stable as "what it isn't", based on a property of the release tag, rather than what "stable" means.

Did you have a specific wording in mind for this?

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 9 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:25]: There is a dedicated paragraph below: `## What is the level of support?`.
Isn't that clear?

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 8 above)
Comment (asklar) [reply proposals/0007-rn-versions-lifecycle-support.md:17]: What I'd like to see happen here is some guarantee along the lines of e.g.: "there are no ABI breaks or other breaking changes within a stable release".

**Review by asklar — CHANGES_REQUESTED** [review]
marking request changes: some minor editorial feedback to make this easier to read, and some open questions re:  support/how security issues will be triaged

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:60]: > however, it does not detail what counts as "bad enough" for a security issue to be worth fixing as per this policy. IMO there should be some criteria regarding what constitutes "we'll fix and backport this issue" vs. "we'll only fix this in latest", vs. "we won't fix it".

I believe the agreement is that we will consider each security issue on a case by case basis, hence the lack of clarity. We can mention this if it makes the policy clearer

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 8 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:17]: > what is the stability guarantee being made?

tl;dr: There are no stability guarantees between versions (this was never the case). Breaking changes are mentioned inside the release notes of each releases in the "Breaking" section. We're not moving to full semver in this policy.

To clarify: the scope of this policy is just to write down what we've been informally doing in the recent past. Should we move versioning or provide further guarantess between versions, we're going to update the policy to mention it.

Comment (asklar) [reply proposals/0007-rn-versions-lifecycle-support.md:17]: understood - can you add a sentence to that effect, since this document talks about what the different release tags mean? e.g. "Note that despite the `-stable` tag, there are no stability guarantees. Breaking changes will be announced in each release's notes."

Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:17]: I would refrain from talking about stability in this document as it's out of scope. If `stable` is not a good term here we can maybe find a better term which is a synonym of `-stable`?

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:60]: Thank you for raising this @asklar  will go through your suggestion and merge them forward in a separate PR.

Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:60]: Make sense @asklar will change this

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 12 above)
Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:26]: Makes sense @asklar will change this as well

**Comment by Andjeliko** [comment]
Thank you for your suggestion @asklar applying your suggestions in a separate PR which we will fast merge

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 12 above)
Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:26]: Makes sense made chagne

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 14 above)
Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:5]: Makes sense made the change

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 10 above)
Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:60]: Made the change

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+|<=0.66.3	|Old minor series	|Unsupported	|
+
+## What is the level of support?
+
+Due to support bandwidth, the React Native team, with the community's help, is looking into issues & PRs opened against one of the supported versions.
+
+Issues & PRs opened against older versions would be considered only in exceptional cases. Please update your application to one of the supported versions and raise the Issue/PR targeted to that version.
+
+Issues should contain a [**reproducer**](https://stackoverflow.com/help/minimal-reproducible-example) project regardless of which version they targeting, for them to be considered.
+Issues without a reproducer will require more effort to understand and fix, and are less likely to receive attention.
```
[the remark this answers is not in the document]
Comment (Andjeliko) [reply proposals/0007-rn-versions-lifecycle-support.md:46]: Made the change

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 5 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:6]: Handed in https://github.com/react-native-community/discussions-and-proposals/pull/513

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 3 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md]: Included in https://github.com/react-native-community/discussions-and-proposals/pull/513

File: proposals/0007-rn-versions-lifecycle-support.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+React Native team will work on best-effort basis to release a new version of React Native every four months. And will focus on speeding up the release cycle and put considerable effort into ensuring a seamless upgrade path.
+This document describes the level of support we can provide and which set of versions is supported.
+
+## Glossary
+
+Below we are defining terms used in this document:
+
+* **stable** **version** - Any version that doesn’t have a -alpha, -beta, -RC postfix in the version name.
+* **latest version** - **** The latest stable version with the highest version number (e.g. 0.68.2).
+    * The highest version number is defined following the [NPM server algorithm](https://github.com/npm/node-semver).
```
[the remark this answers is not in the document]
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:19]: Handled in https://github.com/react-native-community/discussions-and-proposals/pull/513

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 13 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:10]: It applies to any upgade. We're working on upgrading the upgrade path in general 👍

File: proposals/0007-rn-versions-lifecycle-support.md (hunk 8 above)
Comment (cortinico) [reply proposals/0007-rn-versions-lifecycle-support.md:17]: I cross checked other documents and we always refer to `stable`:
https://github.com/facebook/react-native/wiki/Deprecations

I couldn't find a better name and I think that `stable` here is clear enough.

I would also not add the sentence:
"Note that despite the -stable tag, there are no stability guarantees. Breaking changes will be announced in each release's notes."

as this goes in the direction of describing our API evolves, which should be moved to a separate document.

**Comment by cortinico** [comment]
Hey all,
I've handled all your open comments in a separate PR here: 
- #513

We're still accepting comments this week till Mon 19th so please share them on #513 if you have any.
